modules/gdscript/gdscript_compiler.cpp

```diff
@@ -1165,8 +1165,18 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 				bool has_operation = assignment->operation != GDScriptParser::AssignmentNode::OP_NONE;
 				if (has_operation) {
 					// Perform operation.
-					GDScriptCodeGenerator::Address op_result = codegen.add_temporary(_gdtype_from_datatype(assignment->get_datatype(), codegen.script));
 					GDScriptCodeGenerator::Address og_value = _parse_expression(codegen, r_error, assignment->assignee);
+					GDScriptDataType assignment_type = _gdtype_from_datatype(assignment->get_datatype(), codegen.script);
+					if (!has_setter && !assignment->use_conversion_assign && assignment_type.kind == GDScriptDataType::BUILTIN && og_value.type.kind == GDScriptDataType::BUILTIN && assignment_type.builtin_type == og_value.type.builtin_type) {
+						// If there's nothing special about the assignment, perform the assignment as part of the operator
+						gen->write_binary_operator(target, assignment->variant_op, og_value, assigned_value);
```

Comment:
This will not work with static variables after #77129, since the `member` and `target` addresses are not initialized in this case (static variables are currently implemented using separate opcodes, not addressing modes).

```gdscript
static var a: Array = [1, 2]
static var b: Array = [3, 4]

func _run():
    a += b
    print(a)
```

```
  modules/gdscript/gdscript_byte_codegen.h:195 - Leaving block with non-zero temporary variables: 1
  modules/gdscript/gdscript_byte_codegen.cpp:187 - Non-zero temporary variables at end of function: 1
```

Reply:
What is the correct way to detect this scenario? `is_static` doesn't seem to cover it.

Reply:
The `&& !is_static` check should be enough to narrow it down to the current behavior. This means that static variables will not have this optimization.

But we also need to make sure that the code always pops all temporaries before the early return, just like in the code below. [link](https://github.com/godotengine/godot/blob/768c6564dc570603111cf50148395fdb5fdb06a9/modules/gdscript/gdscript_compiler.cpp#L1206-L1214)

I think you need to check `target`. While this seems to work, I think the case with static variables has shown that there is probably a bug with temporaries in this change. I can't tell which one, because the branch above is huge and complex.

`to_assign` looks unnecessary to check at this point, as it is assigned below.